specs/cross-net-messaging.md

```diff
@@ -0,0 +1,279 @@
+
+- 💡 This assumes that contracts need to be implemented to explicitly be able to unwrap and handle general-purpose cross-net messages. We can use for instance `abi.encode("IpcEntrypoint(IpcEnvelope)")` as the entrypoint that contracts need to expose where they handle the execution of general-purpose cross-net messages. In the future, we should provide a library or a SDK to make it easy for contracts to support IPC messages.
+
+In order to simplify the implementation of the `IpcEntrypoint` function, we will add a new `kind` field in `IpcEnvelope` used to clearly identify the type of cross-net message being propagated.
+
+- Both, `Call` and `Receipt` cross-net message types are propagated to the destination contract by calling the `IpcEntrypoint` with the full `IpcEnvelope` as an argument.
+    
+    ```solidity
+    enum IpcMsgKind {
+        Transfer, // use for fund, release and cross-net message that move funds (i.e. no data, so not a smart contract call
+        Call, // for general-purpose messages
+        Receipt,  // for ack messages
+    }
+    
+    struct IpcEnvelope {
+        IpcMsgKind kind;
+    		// abi.encode of the underlying IpcMsgKind 
+    		bytes message;
+    		// outgoing nonce for the envelope.
+    		// this is set by the gateway when committing the message for propagation.
+        uint64 nonce;
+        // fee and gas limit were used to pay for the execution
+        // of messages in destination. Not currently used for the
+        // the first implementation where the gas model has been descoped.
+        uint256 fee;
+        uint256 gasLimit;
+    }
+    
+    // Used for both Call and Transfer messages
+    struct IpcMsg {
+        IPCAddress from;
+        IPCAddress to;
+        uint256 value;
+        bytes4 method;
+        bytes params;
+    }
+    
+    struct ReceiptMsg {
+        IPCAddress to;
+    		bool success;   // we can optionally use an int flag instead of bool
+        bytes32 id;     // id of the message being ack'ed. This id is the keccak of IpcMsg
+    	  bytes reason;   // byte encoded failure reason (as in try-catch)
+    }
+    ```
+    
+- The `id` of an `Receipt` message is determined through the `keccak(IpcEnvelope)` of the message being acknowledged.
+    - `sendXnetMessage` should be extended to return the keccak of the message, for the contract to retain and correlate the result with later
+- Only the execution of `Call` messages trigger `Receipt`, for plain `Transfer` messages, i.e. `fund` and `release` , no `Ack` are needed (at least for now that we are subsidizing these).
```

Comment:
I think if the `Transfer` fails, an `Ack` is also sent?

Reply:
I agree. That line looks out of date to me. I shall update it. Thanks for the comment